presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -1118,6 +1121,30 @@ public PlanWithProperties visitUnion(UnionNode node, Context context)
                 }
             }
 
+            // parent does not have preference or prefers some partitioning without any explicit partitioning - just use
+            // children partitioning and don't GATHER partitioned inputs
+            if (!parentGlobal.isPresent() || parentGlobal.get().isDistributed()) {
+                if (!partitionedChildren.isEmpty() && unpartitionedChildren.isEmpty()) {
+                    if (!hasMultipleSources(plannedChildren.stream().map(PlanWithProperties::getNode).toArray(PlanNode[]::new))) {
```

Comment:
Maybe it's a little bit too late for me, but does your version do the same thing as mine? I want to check whether among all of the children there is at most one table scan. IMO your version perform such check but per each children separately.